File: VB/XtraReport1.vb
Imports Microsoft.VisualBasic
Imports System
Imports DevExpress.XtraReports.UI
Imports System.Globalization
' ...

Namespace CustomGrouping
	Partial Public Class XtraReport1
		Inherits DevExpress.XtraReports.UI.XtraReport
		Public Sub New()
			InitializeComponent()
		End Sub

		Private year As Integer = 0
		Private month As Integer = 0
		Private formatInfo As DateTimeFormatInfo = DateTimeFormatInfo.CurrentInfo

		Private Sub GroupHeader1_BeforePrint(ByVal sender As Object, ByVal e As System.Drawing.Printing.PrintEventArgs) Handles GroupHeader1.BeforePrint
			e.Cancel = month = Convert.ToDateTime(GetCurrentColumnValue("OrderDate")).Month
		End Sub

		Private Sub lbMonthName_BeforePrint(ByVal sender As Object, ByVal e As System.Drawing.Printing.PrintEventArgs) Handles lbMonthName.BeforePrint
			month = Convert.ToDateTime(GetCurrentColumnValue("OrderDate")).Month
			CType(sender, XRLabel).Text = formatInfo.MonthNames(month - 1)
		End Sub

		Private Sub GroupHeader2_BeforePrint(ByVal sender As Object, ByVal e As System.Drawing.Printing.PrintEventArgs) Handles GroupHeader2.BeforePrint
			e.Cancel = year = Convert.ToDateTime(GetCurrentColumnValue("OrderDate")).Year
		End Sub

		Private Sub lbYear_BeforePrint(ByVal sender As Object, ByVal e As System.Drawing.Printing.PrintEventArgs) Handles lbYear.BeforePrint
			year = Convert.ToDateTime(GetCurrentColumnValue("OrderDate")).Year
			CType(sender, XRLabel).Text = year.ToString()
		End Sub

	End Class
End Namespace

File: CS/XtraReport1.cs
using System;
using DevExpress.XtraReports.UI;
using System.Globalization;
// ...

namespace CustomGrouping {
    public partial class XtraReport1 : DevExpress.XtraReports.UI.XtraReport {
        public XtraReport1() {
            InitializeComponent();
        }

        private int year = 0;
        private int month = 0;
        private DateTimeFormatInfo formatInfo = DateTimeFormatInfo.CurrentInfo;

        private void GroupHeader1_BeforePrint(object sender, System.Drawing.Printing.PrintEventArgs e) {
            e.Cancel = month == Convert.ToDateTime(GetCurrentColumnValue("OrderDate")).Month;
        }

        private void lbMonthName_BeforePrint(object sender, System.Drawing.Printing.PrintEventArgs e) {
            month = Convert.ToDateTime(GetCurrentColumnValue("OrderDate")).Month;
            ((XRLabel)sender).Text = formatInfo.MonthNames[month - 1];
        }

        private void GroupHeader2_BeforePrint(object sender, System.Drawing.Printing.PrintEventArgs e) {
            e.Cancel = year == Convert.ToDateTime(GetCurrentColumnValue("OrderDate")).Year;
        }

        private void lbYear_BeforePrint(object sender, System.Drawing.Printing.PrintEventArgs e) {
            year = Convert.ToDateTime(GetCurrentColumnValue("OrderDate")).Year;
            ((XRLabel)sender).Text = year.ToString();
        }

    }
}
